library/src/main/scala/scala/scalajs/runtime/RuntimeLong.scala

```diff
@@ -55,531 +121,558 @@ final class RuntimeLong(
-    var p1 = a1 * b0; // << 13
-    var p2 = a2 * b0; // << 26
-    var p3 = a3 * b0; // << 39
-    var p4 = a4 * b0; // << 52
-
-    if (b1 != 0) {
-      p1 += a0 * b1;
-      p2 += a1 * b1;
-      p3 += a2 * b1;
-      p4 += a3 * b1;
-    }
-    if (b2 != 0) {
-      p2 += a0 * b2;
-      p3 += a1 * b2;
-      p4 += a2 * b2;
-    }
-    if (b3 != 0) {
-      p3 += a0 * b3;
-      p4 += a1 * b3;
+  private def inline_-(alo: Int, ahi: Int, blo: Int, bhi: Int): (Int, Int) = {
+    val lo = alo - blo
+    (lo, ahi - bhi + (if (inlineUnsignedInt_>(lo, alo)) -1 else 0))
+  }
+
+  def *(b: RuntimeLong): RuntimeLong = {
+    val a0 = a.lo & 0xFFFF
+    val a1 = a.lo >>> 16
+    val a2 = a.hi & 0xFFFF
+    val a3 = a.hi >>> 16
+    val b0 = b.lo & 0xFFFF
+    val b1 = b.lo >>> 16
+    val b2 = b.hi & 0xFFFF
+    val b3 = b.hi >>> 16
+
+    var c0 = a0 * b0
+
+    var c1 = c0 >>> 16
+    c1 = c1 + a1 * b0
+
+    var c2 = c1 >>> 16
+    c1 = (c1 & 0xFFFF) + a0 * b1
+    c2 = c2 + (c1 >>> 16)
+
+    var c3 = c2 >>> 16
+    c2 = (c2 & 0xFFFF) + a2 * b0
+    c3 = c3 + (c2 >>> 16)
+    c2 = (c2 & 0xFFFF) + a1 * b1
+    c3 = c3 + (c2 >>> 16)
+    c2 = (c2 & 0xFFFF) + a0 * b2
+    c3 = c3 + (c2 >>> 16)
+    c3 = c3 + a3 * b0 + a2 * b1 + a1 * b2 + a0 * b3
+
+    new RuntimeLong(
+        (c0 & 0xFFFF) | (c1 << 16),
+        (c2 & 0xFFFF) | (c3 << 16))
+  }
+
+  def /(b: RuntimeLong): RuntimeLong = {
+    val alo = a.lo
+    val ahi = a.hi
+    val blo = b.lo
+    val bhi = b.hi
+
+    if (isZero(blo, bhi))
+      throw new ArithmeticException("/ by zero")
+
+    if (isInt32(alo, ahi)) {
+      if (isInt32(blo, bhi)) {
+        if (alo == Int.MinValue && blo == -1) new RuntimeLong(Int.MinValue, 0)
+        else new RuntimeLong(alo / blo)
+      } else {
+        // Either a == Int.MinValue && b == (Int.MaxValue + 1), or (abs(b) > abs(a))
+        if (alo == Int.MinValue && (blo == 0x80000000 && bhi == 0)) MinusOne
+        else Zero // because abs(b) > abs(a)
+      }
+    } else {
+      val (aNeg, aAbsLo, aAbsHi) = inline_abs(alo, ahi)
+      val (bNeg, bAbsLo, bAbsHi) = inline_abs(blo, bhi)
+      val absR = unsigned_/(aAbsLo, aAbsHi, bAbsLo, bAbsHi)
+      if (aNeg == bNeg) absR
+      else inlineLongUnary_-(absR.lo, absR.hi)
     }
-    if (b4 != 0) {
-      p4 += a0 * b4;
+  }
+
+  private def unsigned_/(alo: Int, ahi: Int, blo: Int, bhi: Int): RuntimeLong = {
+    if ((ahi & 0xffe00000) == 0) {
+      if ((bhi & 0xffe0000) == 0) {
+        val aDouble = ahi * TwoPow32 + alo.toUint
+        val bDouble = bhi * TwoPow32 + blo.toUint
+        val rDouble = aDouble / bDouble
+        new RuntimeLong(rawToInt(rDouble), rawToInt(rDouble / TwoPow32))
+      } else {
+        Zero // because b > a
+      }
+    } else {
+      if (bhi == 0 && isPowerOfTwo_IKnowItsNot0(blo)) {
+        val pow = log2OfPowerOfTwo(blo)
+        if (pow == 0) new RuntimeLong(alo, ahi)
+        else new RuntimeLong((alo >>> pow) | (ahi << -pow), ahi >>> pow)
```

Comment:
`inline_>>>` would have an additional check for `pow >= 32`, which we know is not true here, but we have no well to give that information to the optimizer so that it will eliminate it.